drivers: sensor: qdec_stm32: add optional encoder mode

jeppenodgaard opened the pull request:
Add support for setting encoder mode in dts.

Commit:
drivers: sensor: qdec_stm32: add optional encoder mode

Add support for setting encoder mode in dts.

Signed-off-by: Jeppe Odgaard <jeppe.odgaard@prevas.dk>

jeppenodgaard commented:
I used the qdec sample to test single phase encoder on a nucleo_h563zi.
I can add the overlay in a commit if it adds value.

erwango commented on dts/bindings/sensor/st,stm32-qdec.yaml:
```diff
@@ -16,6 +16,13 @@ properties:
+  st,encoder-mode:
+    type: int
+    description: |
+      Set encoder mode. The value depends on the MCU series. A description of
+      the values can be found in the reference manual or the HAL header. If not
+      set a default value is chosen by the HAL encoder driver.
```
Several issues with this description:
Values should be documented. If there is no ambiguity by reading the RM, it should be doable to document them.
If values depends on series, then you might need to update the code to allow a single set of values that are then properly encoded by the driver.
For default value, be more explicit, at least we should get a sense of the functional impact.

See ` st,input-filter-level` as a good example.
See also that, when doable, providing an enum allows early dts level checks.

erwango reviewed (changes requested):
Thanks for this contribution, some rework required about the way to encode the property.

Thanks for the proposal for overlay, but since we miss proper sample so I don't this it could be useful.

Commit:
drivers: sensor: qdec_stm32: clang-format qdec_stm32.c

Run clang-format on qdec_stm32.c.

This moves '\' in macros to line index 100, which ensures '\' still
aligns if a macro is edited with a new line that is longer than the
previous longest line.

Signed-off-by: Jeppe Odgaard <jeppe.odgaard@prevas.dk>

jeppenodgaard commented:
Thanks for the review @erwango !
I've split the commit into two commits.
I tried to address the issues by comparing f401 and h563 and make common documentation and values based on those RMs.

erwango commented:
> @erwango any chance you got time to review the new changes?

You didn't pushed the new changes.

jeppenodgaard commented:
> > @erwango any chance you got time to review the new changes?
> 
> You didn't pushed the new changes.

My bad. Pushed them now.

erwango commented on dts/bindings/sensor/st,stm32-qdec.yaml:
```diff
@@ -16,6 +16,34 @@ properties:
+      0x10006: Quadrature encoder mode: x1 mode, counting on tim_ti1fp1
+      0x10007: Quadrature encoder mode: x1 mode, counting on tim_ti2fp2
+
+      If unset encoder mode is defined by LL_TIM_ENCODER_StructInit().
```
Why not using a default value which will be matching LL default? See https://docs.zephyrproject.org/latest/build/dts/bindings-syntax.html#default

Also, bindings should not reference code (they are supposed to be 'implementation agnostic' ) but refer to behavior. So you should rather document "what will be the effect when not set".

Note, if there are 2 types of encoding IPs which have different behavior which makes it complex to document with a single description, instantiate a second compatible where your can document a specific behavior.

jeppenodgaard replied:
Good question.
I went through the HAL and all have the same default value.
I've added default which shouldn't break backwards compatibility.

Thanks for the additional bindings insight :+1:

erwango reviewed (changes requested):
Thanks for the changes, this is much easier to review this way.

erwango commented on drivers/sensor/qdec_stm32/qdec_stm32.c:
```diff
@@ -138,13 +147,23 @@ static const struct sensor_driver_api qdec_stm32_driver_api = {
 		.pclken = {.bus = DT_CLOCKS_CELL(DT_INST_PARENT(n), bus),                          \
 			   .enr = DT_CLOCKS_CELL(DT_INST_PARENT(n), bits)},                        \
+		.encoder_mode =                                                                    \
+			DT_INST_PROP_OR(n, st_encoder_mode, QDEC_STM32_ENCODER_MODE_UNSET),        \
```
Since you now provide a default value, property will always be there and `DT_INST_PROP_OR` and `QDEC_STM32_ENCODER_MODE_UNSET` related code it no more required

jeppenodgaard replied:
Right. Should be fixed now.

erwango reviewed (dismissed):
Thanks for the changes

MaureenHelm commented on dts/bindings/sensor/st,stm32-qdec.yaml:
```diff
@@ -16,6 +16,33 @@ properties:
+      0x10005: Encoder mode: Directional Clock, x1 mode
+      0x10006: Quadrature encoder mode: x1 mode, counting on tim_ti1fp1
+      0x10007: Quadrature encoder mode: x1 mode, counting on tim_ti2fp2
+    default: 0x1
```
This needs a justification. Is it a register reset value? Please document in the description why the default value was chosen

jeppenodgaard replied:
Thanks. I added default value documentation.

erwango replied:
Sorry, but one last comment on this part:
This will render as follows in doc: https://docs.zephyrproject.org/latest/build/dts/api/bindings/sensor/st%2Cstm32-qdec.html#dtbinding-st-stm32-qdec
As you can see, the description part is displayed, but not the default. So I, d' suggest to do something like:
```
    description: |
      Set encoder mode.
      The following modes are supported:
      0x1: Encoder mode 1 (Default)
      0x2: Encoder mode 2
      0x3: Encoder mode 3
 [...]
```

erwango reviewed (changes requested):
One very last change and this will be fine. Thanks for the changes so far

jeppenodgaard commented:
> One very last change and this will be fine. Thanks for the changes so far

Makes sense. Thanks :)
